### ISSUE TITLE
qemu_cortex_m0: Integrate nRF51 timer patch

### ISSUE THREAD

**Issue by microbuilder** [zephyrproject-rtos/zephyr#58990]
The timer used by `qemu_cortex_m0` is based on the `microbit` machine in QEMU, which relies on an emulation of the nRF51.

The main clock works as expected when no `icount` value is set, but exhibits problems when a value is set for `icount`.

This is caused by a bug in the underlying nRF51 timer emulation code in QEMU, as described by @pm215 here: https://github.com/zephyrproject-rtos/zephyr/pull/58753#issuecomment-1578735047

The following QEMU patch may resolve this issue (testing required): https://patchew.org/QEMU/20230606134917.3782215-1-peter.maydell@linaro.org/

This bug report is being raise to track integrating this patch into QEMU in the Zephyr SDK, or updating to a version of QEMU that includes this patch next release, or report of testing issues encoutered with the proposed patch.

CC @carlocaione @stephanosio @carlescufi 

**To Reproduce**

With: `west build -b qemu_cortex_m0 samples/hello_world -p` and
`4c8b18920a7c1393c3f182af85a4232b7280f6d1` this diff can be used:

```diff
diff --git a/samples/hello_world/src/main.c b/samples/hello_world/src/main.c
index bc7a4c1d71..1cfcdfe456 100644
--- a/samples/hello_world/src/main.c
+++ b/samples/hello_world/src/main.c
@@ -8,6 +8,11 @@
 
 int main(void)
 {
-       printk("Hello World! %s\n", CONFIG_BOARD);
+       unsigned int key = irq_lock();
+       k_busy_wait(15000);
+       irq_unlock(key);
+
+       printk("done\n");
+
        return 0;
 }
```

Using QEMU emulator version 7.2.0 (Debian 1:7.2+dfsg-5ubuntu2) I get:

    working version:

```bash
qemu-system-arm -cpu cortex-m0 -machine microbit -nographic -kernel build/zephyr/zephyr.elf -rtc clock=vm
```

    non-working version:

```bash
qemu-system-arm -cpu cortex-m0 -machine microbit -nographic -kernel build/zephyr/zephyr.elf -rtc clock=vm -icount shift=6,align=off,sleep=off
```

**Comment by pm215** [comment]
FWIW, the fix is in QEMU 7.2.4, 8.0.3, and will be in the upcoming 8.1.

**Comment by microbuilder** [comment]
> FWIW, the fix is in QEMU 7.2.4, 8.0.3, and will be in the upcoming 8.1.

Thanks for the update. I'll close this when we're on an appropriate version in the Zephyr SDK.

**Comment by pm215** [comment]
Do you plan to also reenable CONFIG_QEMU_ICOUNT when you have a QEMU that will let you do it?

**Comment by microbuilder** [comment]
> Do you plan to also reenable CONFIG_QEMU_ICOUNT when you have a QEMU that will let you do it?

We generally only enable ICOUNT as a workaround when it's required and we have timing problems. I'll have to test with the version of QEMU that will go into the SDK.

**Comment by keith-packard** [comment]
I'm seeing consistent failures in tests/kernel/interrupt/arch.interrupt when icount is not being used. Is it likely that the test is broken? Or does it just need to force icount to be used?

**Comment by microbuilder** [comment]
@keith-packard We can re-enable ICOUNT by default, but to fix issues with the Zephyr timer used by this QEMU machine target we'll require a new Zephyr SDK to be released based on QEMU 8.0.3 or 8.1+ (which is at RC4 as of a week or so ago).
